pipettes/tests/test_motor_pulse.cpp

```diff
@@ -0,0 +1,136 @@
+#include <stdio.h>
+
+#include <iostream>
+
+#include "catch2/catch.hpp"
+#include "motor-control/core/motor_interrupt_handler.hpp"
+#include "pipettes/tests/mock_message_queue.hpp"
+
+using namespace motor_handler;
+
+TEST_CASE("move a full rotation") {
+    MotorInterruptHandler<mock_message_queue::MockMessageQueue> handler{};
+    mock_message_queue::MockMessageQueue<Move> queue;
+    handler.set_message_queue(&queue);
+    handler.reset();
+
+    GIVEN("in full step mode") {
+        // one rotation in full stepping = 200 steps
+        sq32_31 temp_steps = 0xC8LL << 31;
+        Move msg1 = Move{.target_position = temp_steps};
+        queue.try_write(msg1);
+        REQUIRE(handler.get_current_position() == 0);
+        bool running = true;
+        while (running) {
+            handler.pulse();
+            if (!handler.has_active_move) {
+                running = false;
+            }
+        }
+        REQUIRE(handler.get_current_position() == temp_steps);
+    };
+    GIVEN("in 256th micro-stepping mode") {
+        // one rotation in 256th microstepping = 51200 steps
+        sq32_31 temp_steps = 0xC800LL << 31;
+        Move msg1 = Move{.target_position = temp_steps};
+        queue.try_write(msg1);
+        REQUIRE(handler.get_current_position() == 0);
+        bool running = true;
+        while (running) {
+            handler.pulse();
+            if (!handler.has_active_move) {
+                running = false;
+            }
+        }
+        REQUIRE(handler.get_current_position() == temp_steps);
+    };
+}
+
+TEST_CASE("move less than a full step") {
+    MotorInterruptHandler<mock_message_queue::MockMessageQueue> handler{};
+    mock_message_queue::MockMessageQueue<Move> queue;
+    handler.set_message_queue(&queue);
+
+    GIVEN("one fractional move") {
+        // equivalent of 28.23 steps
+        sq32_31 temp_steps = 2823LL << 29;
+        Move msg1 = Move{.target_position = temp_steps};
+        queue.try_write(msg1);
+        REQUIRE(handler.get_current_position() == 0);
+        bool running = true;
+        while (running) {
+            handler.pulse();
+            if (!handler.has_active_move) {
+                running = false;
+            }
+        }
+        REQUIRE(handler.get_current_position() == temp_steps);
+    }
+
+    GIVEN("two fractional moves") {
+        // equivalent of 28.23 steps
+        sq32_31 temp_steps = 2823LL << 29;
+        Move msg1 = Move{.target_position = temp_steps};
+        queue.try_write(msg1);
+        queue.try_write(msg1);
+        REQUIRE(handler.get_current_position() == 0);
+        bool running = true;
+        while (running) {
+            handler.pulse();
+            if (!handler.has_active_move && !handler.has_messages()) {
```

Comment:
It's unfortunate that it's possible to make this test get into an infinite loop rather than fail.